### ISSUE TITLE
ctx.fetch() not keeping original headers

### ISSUE THREAD

**Issue by aquerrec** [mswjs/msw#145]
## Describe the bug
When mocking a REST API using `rest.get` i have a request containing headers, specifically "Authorization" header, and when i call `ctx.fetch()` with this request, in order to get the original response, i have an unauthorized response.
This is due to the fact that my "Authorization" header is missing.

After taking a look at `src/context/fetch.ts`, specifically the `augmentRequestInit()` function, it seems that the type of my request headers is `Headers`, and not `string[][]` nor `Record<string, string>` (see `HeadersInit`type).
I think that the spread operator is not working for the type `Headers`, therefore the headers are not kept :
```typescript
const augmentRequestInit = (requestInit: RequestInit): RequestInit => {
  return {
    ...requestInit,
    headers: {
      ...requestInit.headers,
      'x-msw-bypass': 'true',
    },
  }
}
```

## Environment

- `msw: 0.15.5`
- `nodejs: 12.16.1`
- `npm: 6.13.4`

- browser version: Chrome `81.0.4044.129`

## To Reproduce
1. make a request with "Authorization" header in an authent context
2. call `ctx.fetch()` function with the original request
3. you will see an "401 Unauthorized (from ServiceWorker)" request in your network

My mock :
```typescript
rest.get('/path/to/my/resource', async (req, res, ctx) => {
        const mockedResponse = [
            "Val 1",
            "Val 2",
            "Val 3"
        ];

        const originalResponse = await ctx.fetch(req)

        return res(
            ctx.status(200),
            ctx.json(originalResponse && originalResponse.length ? originalResponse : mockedResponse)
        );
    })
```

## Expected behavior
Headers are kept.

## Temporary Solution
I implemented my own `fetch()` duplicating the existing one, and i change the implementation of the `augmentRequestInit()` function.
My working version :
```typescript
const augmentRequestInit = (requestInit: RequestInit): RequestInit => {
    const headersCopy: Record<string, string> = {}
    if (requestInit.headers instanceof Headers) {
        requestInit.headers.forEach((value: string, name: string) => {
            headersCopy[name] = value
        });
    }
    
    // TODO handle the other types : string[][] and Record<string, string>

    return {
        ...requestInit,
        headers: {
            ...headersCopy,
            'x-msw-bypass': 'true',
        },
    }
}
```

**Comment by kettanaito** [comment]
Hey, @aquerrec. Thanks for reporting this. You're right, it seems that the original headers cannot be spread like this, being an instance of `Headers`. 

I believe we can construct the `Headers` instance and use it as the value of the `headers` key. In the `context/fetch.ts` file we need to change the `augmentRequestInit` function:

https://github.com/mswjs/msw/blob/d812597b238cb95b99453161dd30e9a1bd11ae5d/src/context/fetch.ts#L9-L17

To construct the `Headers` instance, instead of handling it as an object:

```js
// context/fetch.ts
const augmentRequestInit = (requestInit: RequestInit): RequestInit => {
  // Construct the Headers instance
  const headers = new Headers(requestInit.headers)
  headers.set('x-msw-bypass', 'true')

  return {
    ...requestInit,
    headers,
  }
}
```

> The only doubt I have about this is whether you can construct `Headers` instance given another `Headers` instance.

Would you be interested in helping me out with this change? I'll support you with integration tests, so we are sure this works as intended. Please let me know.

**Comment by aquerrec** [comment]
Hey, @kettanaito, your fix works like a charm !
But i have not tested with `string[][]` nor `Record<string>`, only with `Headers`.
Thanks

**Comment by juhanakristian** [comment]
Hi @kettanaito I would like to help with this but I'm not quite sure how to start with the integration tests. 

Should there also be unit tests for `augmentRequestInit`?

**Comment by kettanaito** [comment]
@juhanakristian, hey. You're right, a unit test for `augmentRequestInit` is a good idea.
I can help you with the integration tests once you provision a pull request. Sounds fair?

**Comment by juhanakristian** [comment]
Yeah, that sounds great!

**Comment by kettanaito** [comment]
Big thanks to @juhanakristian for providing a fix on this.

The fix is published in [`0.15.8`](https://github.com/mswjs/msw/releases/tag/v0.15.8). Could you please update and let me know that it functions as expected? The workaround you've introduced should no longer be needed, original request headers must be proxied via `ctx.fetch()`. Thanks.
